### DISCUSSION TITLE
Should all html tags be persisted?

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#1231] [Announcements]
Recently I found that html utilities used by Freeplane can produce garbage resulting in invalid xml if user paste html from external sources that contain unknown tags, and the set of the known tags is a subset of html 4.

There are two solutions:

1. removing all unknown tags because they are not interpreted and have no influence on how the content looks,
2. using external library which can try to repair the xml.

In my eyes Freeplane, should not try to preserve unknown tags because its only purpose is showing and searching the mind map content. Any unknown tags not interpreted by Freeplane are garbage the most users can neither process nor are interested in. Therefore, removing such tags like "circle", "g" and so on is the right way. It make maps smaller, Freeplane faster and safer.

However, in our discussion in issue https://github.com/freeplane/freeplane/issues/1225 one particular user expressed an interest to keep this content. For doing so I would have to add another library to Freeplane which can repair the content.

So I would like to ask if you saving "unknown" tags is important for your use cases and if Freeplane should support it, should I introduce an option, which default I should choose and why.

I am interested to get quick responses because this decision is relevant also for releasing urgent bug fixes. 

Regards,
Dimitry

**Comment by steveatmaat** [comment]
Unknown tags are of no use to me, and I think they should be removed. I use FreePlane mostly as a knowledge base, and I would prefer to keep it lightweight and fast.

**Comment by macmarrum** [comment]
If Freeplane was considered a closed system, I would vote for removing any additional tags beyond what is supported by Freeplane itself.
Yet Freeplane can be and often is used as a kind of database, on top of which some processes are run, gathering information from nodes, modifying it and saving. E.g. Markdown nodes can have tags that aren't supported by Freeplane but after exporting, they can be interpreted and are useful.
Therefore, I am of the opinion that Freeplane should allow to keep extra tags (if possible), preferably as an option.

**Reply by EdoFro** [reply]
With a "Safelist" .cfg file in the userdirectory that can be modified by the user if he/she wants to add other tags, etc.

**Reply by dpolivaev** [reply]
I am implementing it because I trust your judgment which is actually strongly different from mine. In my eyes, the vast majority of the users would never be interested in use cases requiring the additional tags to be kept, but now I should bother all of them with making the decision. But it is what we in society call "minority rights".

**Comment by dpolivaev** [comment]
I am introducing a new option:

OptionPanel.saveCompactHtml=Save Compact HTML
OptionPanel.saveCompactHtml.tooltip=Same look, less code. Saves space and speeds up loading.

**Comment by Bricklam** [comment]
Thanks for your work @dpolivaev.
by the way, I no longer insist on this issue and it is totally OK if you want to keep the new changes (removing the unknown tags) .
I have found a workaround by scripting to parse & save the entire Html code for any rich text in the clipboard and then I can save it as plain text into the (attributes as you suggested earlier) or in the notes/details as text.
 this way will keep the XML file cleaner and safe from unmatched tags.
anyway, it is great that you decided to add a new option and to give the user a full control over this issue. this will definitely cover more use cases and allow more options and make Freeplane a better application.

**Comment by sifran-github** [comment]
Is it possible to leave the tags as they are, only to omit them from being rendered when freeplane displays the content?

For example, if you have a function clean_tags(), then instead of using it when saving the data, just use it before sending the saved html for rendering.

**Reply by dpolivaev** [reply]
Unfortunately, it is not possible because of multiple reasons.

**Reply by guoranjun2** [reply]
Hi Dimitry, I tested the new version and found that the HTML styles I edit would automatically disappear, is this due to some changes made in the new version?
here is my node html:
![image](https://github.com/freeplane/freeplane/assets/50692082/7efde560-5b03-4028-bd72-47420b4435cf)
after I saving it, it changed to:
![image](https://github.com/freeplane/freeplane/assets/50692082/7ecfbe43-38cd-49b4-99c6-a85625a5e59d)

**Reply by dpolivaev** [reply]
Please submit your code snippet. I am not an OCR machine. I do not like typing it from the picture !!!

**Reply by guoranjun2** [reply]
ok, here is my snippet, you can test it in your machine: 
```
<html>
  <head>
    
  </head>
  <body>
    <p style="font-size: 12pt;">
      hello world
    </p>
  </body>
</html>
```

**Reply by dpolivaev** [reply]
Thanks. I'll test it. In the meanwhile, I have learned how to use my computer to do the OCR for me.

**Reply by guoranjun2** [reply]
congratulation, I also use that a lot too, it is a very useful function.

**Reply by guoranjun2** [reply]
and it only happened in coreText and detailsText

**Reply by dpolivaev** [reply]
I am going to check it later this week

**Reply by guoranjun2** [reply]
sure, have a good day.

**Comment by dpolivaev** [comment]
Preview 1.11.5_09 should fix all mentioned issues.

**Reply by Bricklam** [reply]
well done. I checked the source code and detected the use of external library: "JSoup".

it is a very good option, but sometimes I prefer to use the attributes to save the html source code as a plain text (to keep the original code intact), and it works fine.
by the way I have a new feature requrest related to saving the attribute text, you can check it here:
https://github.com/freeplane/freeplane/issues/1241

Thanks

**Reply by guoranjun2** [reply]
I just tested the new version and the bug has been fixed. Thanks for the work.

**Comment by macmarrum** [comment]
I have tested v1.11.5-pre09 with regard to saving the original tags (with `Preferences…->HTML->Html Import->Save Compact Hypertext` disabled).

The content under test was copied from https://osintframework.com/ in Chromium on Windows 10.

### 1. Copy and paste `Username`

#### Original Transferable
```html
 <span style="color: rgb(0, 0, 0); font-family: &quot;Helvetica Neue&quot;, Helvetica; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: nowrap; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; display: inline !important; float: none;">Username</span>
```

#### HTML in SimptyHTML after pasting
```html
<html>
 <head>
 </head>
 <body>
  <p><font color="rgb(0, 0, 0)" face="Helvetica Neue, Helvetica">Username</font></p>
 </body>
</html>
```

### 2. Copy and paste `Email Address` and `Username`

#### Original Transferable
```html
 <g class="node" transform="translate(180,35.625)" style="color: rgb(0, 0, 0); font-family: &quot;Helvetica Neue&quot;, Helvetica; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><a target="_blank"><text x="-10" dy=".35em" text-anchor="end" style="fill-opacity: 1;">Email Address</text></a></g><g class="node" transform="translate(180,11.875)" style="color: rgb(0, 0, 0); font-family: &quot;Helvetica Neue&quot;, Helvetica; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><circle r="6" style="fill: rgb(176, 196, 222);"></circle><a target="_blank"><text x="-10" dy=".35em" text-anchor="end" style="fill-opacity: 1;">Username</text></a></g>
```

#### HTML in SimptyHTML after pasting
```html
<html>
 <head>
 </head>
 <body>
  <p><a class="node" transform="translate(180,35.625)" target="_blank" style="color: rgb(0, 0, 0); font-family: Helvetica Neue, Helvetica; font-size: 14px; font-style: normal; font-weight: 400; letter-spacing: normal; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; word-spacing: 0px; background-color: rgb(255, 255, 255)"><font color="rgb(0, 0, 0)" face="Helvetica Neue, Helvetica" size="14px"><text x="-10" dy=".35em" text-anchor="end" align="start" style="letter-spacing: normal; text-indent: 0px; text-transform: none; white-space: normal; word-spacing: 0px; background-color: rgb(255, 255, 255)">
      Email Address
     </text></font></a><font color="rgb(0, 0, 0)" face="Helvetica Neue, Helvetica"><g class="node" transform="translate(180,11.875)" align="start" style="letter-spacing: normal; text-indent: 0px; text-transform: none; white-space: normal; word-spacing: 0px; background-color: rgb(255, 255, 255)"></g></font><circle r="6"></circle><a target="_blank"><text x="-10" dy=".35em" text-anchor="end">
     &#xa0;&#xa0;Username
    </text></a></p>
 </body>
</html>
```

Original Transferable was obtained using `script1` on `root`.
[test-FP1231-should-all-html-tags-be-persisted.xi.zip](https://github.com/freeplane/freeplane/files/11779226/test-FP1231-should-all-html-tags-be-persisted.xi.zip)

## Conclusions

The resulting content in SimpyHTML is quite far from the original. I am not sure this functionality will be useful to many people. It is unlikely that I will be using it. 

People might be better off using the `Markdown` format of a node and storing their original XML there (with a space added at the beginning, before `<html>`, to prevent Freeplane from changing it). 


## Final Remarks

Should the functionality be retained, there is a bug to be fixed: 
`Preferences…->HTML->Html Import->Save Compact Hypertext` set to `ASK` acts as `NO`. I'd expect for a dialog to appear.

**Reply by dpolivaev** [reply]
HTML handling in Java Swing is generally extremely complicated.
Preserving the original html code with tags not available in Swing is **generally not possible**.
I myself was against adding this complexity because it is quite an edge case. 
However, people from the community in this thread urged me to find some solution.
I believe that there is generally no better solution than the solution I have developed.
If you think it is not helpful I can remove it so that the unknown tags are never preserved.
It would prevent the users from being asked a question about saving hypertext in compact form which is likely to be unclear to most of them.

Otherwise, I can fix the bug with the "ASK", it is pretty easy.

Please let me know what you want me to do.

@macmarrum  @EdoFro @Bricklam

**Reply by EdoFro** [reply]
Hi @dpolivaev,
I personally don't need the TAGS, that simplyHTML doesn't use, to remain in the text.
My only interest is that when pasting html or xml, this does not corrupt the map file and that the formats that SimpleHTML does accept, do display correctly.

I suggested jSoup because it is the tool I know of, that allows you to sanitize html and xml before they are pasted in Freeplane, to avoid any problems. 

bye,
edo

**Reply by dpolivaev** [reply]
Anyway, thanks to @macmarrum for careful testing and to @EdoFro for sharing jsoup.

**Reply by dpolivaev** [reply]
It looks like no one here is interested in the current solution of preserving the tags. So, if nobody votes for keeping them I'll remove them.

**Reply by macmarrum** [reply]
@dpolivaev Thanks for providing a solution that allows us to see in action how the functionality can work :rocket: 
This truly agile :+1: 

Now that we know, including what limitations there are, we can make an informed decision.

I vote for Freeplane to keep only the tags that it needs.

**Reply by dpolivaev** [reply]
@macmarrum I found one potential improvement which would allow me to keep the original HTML at least until it is edited, but editing would bring it to the form you already know. Should I try it?

**Reply by macmarrum** [reply]
@dpolivaev Thanks for this new idea.

I found out that it's possible to keep the original HTML in a Markdown node (with a space added at the beginning, before `<html>`, to prevent Freeplane from changing it). This should be sufficient for people wanting to preserve the content without the need to edit it.

And if they change their minds and want to have it editable, they can copy-paste the html code to SimplyHTML (HTML Code view) so that it's converted to whatever Freeplane considers useful html.

Therefore, I see no additional value of such a solution.

**Comment by dpolivaev** [comment]
Preview 1.11.5_10

**Comment by Bricklam** [comment]
I also vote for keeping the known tags only, and apologize for my previous request. I didn't realize all the details behind it

I notice that some requests and suggestions made by me recently is caused by the fact that 
I don't fully understand the complexity and the intricacy of Freeplane project and how its components work on a deep level.
since I see things from a user standpoint, not from a developer standpoint.

for example, as a user I wasn't fully familiar of Freeplane capabilities and limitations.
and these limitations are not because of the developers or the development process. but mostly due to the fact that Freeplane was built on Swing, with all its complexity, limitations and shortcomings, which make its development more difficult and time consuming.

 However, this fact didn't prevent Freeplane from being a great and amazing application that I consider it to be the best mind mapper and information manager. it really exceeded the expectations for a Java Swing application.

that is why I have a lot of appreciation for the main developer @dpolivaev and other who help hem for going all the way to build such a nice application with all these functionalities and features, regardless of all the difficulties and restrictions.

I really want to know more about Freeplane on a deep level and I want to understand all the details and logic behind it.
I hope someday to be able to contribute to its development, I have a good background in Java Programming. and started recently to dive into Swing and I'm doing very well.

The problem when I started to look into Freeplane code: the code base is really complex and confusing and lacks documentation.
(unlike the scripting API, which is fairly clear, documented and easy to understand).

I got lost in the code, didn't know where to start. and I found it very hard to connect the code files to their corresponding functions or components. and I also found difficult to get the general layout or structure behind the code.
and I nocied that the comments are almost absent from the code files, so there is no explanation of what is going on on a deep level.
this makes understanding the code very difficult. However, there was a positive point: the names of variable and classes are very indicative can provide a lot of information, still not enough for me to understand all the details.

that is why I think I need more resources, to help me understand the code structure and architecture
it would be a good idea if you can direct me to any (documentation, charts, UMLs, maps,..) that can help me how to visualize the code as the developer visualizes it.

I just need a visual introduction to the code, and how it is structured and what tools I can use to analyze it.
as an entry point to understanding Freeplane and may encourage me to join the development team

Best regards

**Reply by dpolivaev** [reply]
I started Freeplane as a fork of FreeMind at times I knew almost nothing about clean code principles. As a consequence, Freeplane often follows the open-closed principle but violates many others. It makes it hard for new people to start contributing without guidance on their own.

Therefore when new devs want to join I prefer to analyze their feature ideas together with them and to give them very detailed guidance about possible implementation. Actually, I like onboarding new devs,  thinking together, and pairing. So if you want to contribute it might be a good way to go.

**Reply by Bricklam** [reply]
Thanks very much for this opportunity, I'd be very happy if I was able to implement any useful features or ideas, or to come up with new ones. I'm really interested in any contribution that can improve Freeplane by extending its features or adding new ones. I'm impressed by the pace of development and it is good opportunity for me to be onboard.

I need some time to finish off some personal projects before I can able to devote myself to Freeplane contribution. and during this time I can start by going through all the available documentation &  familiarize myself with the code. So, I can feel more ready & less scared to dive into real work.

by the way I found a very good page in the documentation (old wiki) that layout the main architecture of Freeplane:
[https://docs.freeplane.org/attic/old-mediawiki-content/Architecture.html](https://docs.freeplane.org/attic/old-mediawiki-content/Architecture.html)
[https://www.freeplane.org/oldwiki/index.php/Architecture](https://www.freeplane.org/oldwiki/index.php/Architecture)

this page introduced me to some interesting information about Freeplane code architecture, which is a good start.
also there are some diagrams (code analysis) done by (Stan4j) Java code analyzer, which appears to be a good application.

So, I will go through all available documentation along with my diving into Swing, until feeling some confidence to start contribution.

I would be happy to know if there are any pre-planned features or determined development path for the coming versions, or if these issues are still in the making. any guidance would be appreciated.

Thanks

**Reply by dpolivaev** [reply]
Most people who contributed to Freeplane have implemented their own ideas or ideas they selected themselves. When you know what you are interested to implement let us talk, and if you bring different ideas it would give us some choices. I am confident with the code base and happy to help.

**Comment by Patronos** [comment]
Hello There. I know that I am late to the show but please understand that I found out about this new situation around two days ago. I rarely follow any discussions of most of the programms that I use otherwise it would end up becoming a fulltime job in doing so.

So - I found out the hard way about the removal of not supported html tags after I did my usual tasks by using this great piece of software. I edited my mindmaps and saved them as I usually did. Later on I figured out by accident that the size of my mindmaps decreased by a numerous amout of bytes. This lead me to run a diff -ruN old.mm new.mm on them where I ended up being introduced that a bunch of my html tags have disappeared.

I came here on the mailinglist and was accidently commenting on a thread where I believed, that I was in the right place. There used to be comments about broken html tags and reworking of the selfhtml engine or something like that. My comment can be found here:

https://github.com/freeplane/freeplane/discussions/1230#discussioncomment-6340353

I would like to explain something about my specific use case with freemind and later freeplane. I use this software for many many years now. During that time a not less amount of mindmaps have been created by me. Mindmaps that keep all kind of personal informations inside. Some of the Mindmaps are what they are - Mindmaps. Some of them are bigger with a lot of Notes, Informations and even html tags inside. Some of them even contain tables and html comments.

I have also created scripts (usually bash) that has a specific mindmap as input and generate an output of the html sections surrounded by comments sections, so I can automaticly generate an output for later processing. Some of the mindmaps even have specific colgroup and col tags inside so - whenever I export them - generate a visual pleasing appearence of the resulting web page (let's name it like that) without having to deal with plenty of width attributes inside the html tags.

I was surprised that - after all these years and all of a sudden - the written mindmap dropped all of these tags. Without a warning or a popup that informed people that this is going to happen.

Even the comment tags have been removed as I just tested.

The comment tags would have been at least one small hope that I could wrap necessary tags or sections into comments for parsing and/or further processing.

All gone. I really depend on this. This was exactly one of the major features in freeplane that I really liked for many years. No other minmap solution offers this.

Can I at least re-activate this conversation or at least be one of the users who believes that this really was not an optimal decision. At least please re-introduce the comment tags, that would still generate a bit of work for me to change all my mindmaps, but it will at least be a compromise, so I can wrap all the stuff that may matter to me within comments. Within comments I also mean commented html tags that should *not* be removed :)

Thank you!

**Reply by macmarrum** [reply]
I am not sure I can say anything more about my needs and opinion of this functionality. All was already said in my previous comments.

So maybe I'll just try to describe the history as I see it.
In v1.10.x, the rich-tech editor kept some of additional tags and attributes. Probably not much changed in comparison to earlier branches (v1.9.x, ...).
In v1.11.x, something changed and HTML seemed to be reformatted / rewritten. Some people noticed disappearing tags and/or attributes. This thread was created. People expressed their opinions and Dimitry had a look at what is possible implementation-wise.

There was an attempt to keep additional tags – v1.11.5-pre09, which I tested, The results (in one of my comments above) were less satisfactory than expected.

@Patronos If v1.10.x is working for you, I think you might need to consider sticking with it.

**Reply by Patronos** [reply]
> The results (in one of my comments above) were less satisfactory than expected.

From what I remember (I asked Dimitry, when he was still working on Freemind many years ago, to switch to JavaFX's rendering engine which understands HTML4) SimplyHTML understands HTML3 (at least not all of it if I recall correctly). Therefore the CSS Support might have been limited to maybe CSS and a subset of CSS2 (just speculating). Dimitry for sure knows it much better. I only repeat the parts that I still remember from that conversation and I might be wrong because it was so long ago.

In your tests you've been using CSS styles that goes much further than what SimplyHTML might understand. Therefore you end up getting these results. But if you stick to basic HTML3 and basic CSS styles, then you can get quite nice results. I use the source editor to write simple HTML code inside it.

The errors mentioned initially usually come up if someone open some (invalid or not closed) tags in the source editor section of the "notepad" and then switch to view and then back to source view. This usually causes errorous visual results. At least that's what I observed from own experiences - and yes that's clearly problematic.

@dpolivaev: Thanks for addessing the problem.

**Reply by macmarrum** [reply]
> In your tests you've been using CSS styles that goes much further than what SimplyHTML might understand. Therefore you end up getting these results. But if you stick to basic HTML3 and basic CSS styles, then you can get quite nice results. I use the source editor to write simple HTML code inside it.

Cool. We definitely need a new approach to testing it, if and when it's re-worked by Dimitry. Maybe you could do (some of) the testing?

**Comment by Patronos** [comment]
> @Patronos If v1.10.x is working for you, I think you might need to consider sticking with it.

Well I was asking for a compromise to re-introduce at least the html comment tag.

`<!-- comment -->`

So I can wrap the stuff inside it. Everything else can stay as it is now. Removing everything is not really an option.

E.g.:

```
<!-- colgroup>
  <col ...>
  <col ...>
</colgroup -->
```

I am clearly not the only one who seem to bring this point up. Once this makes its rounds, it might be that more people are bringing this up. And suggesting a long time freemind and freeplane user to stick to an old version (which it sooner or later will become) is really short sighted. So why has this been brought up for open discussion if the direction where it has to go was already clear ? At least from my perception.

The disappearing of tags is equally wrong than having garbaged tags inside the xml structure due to errorous or mistyped tags. I trapped in that issue as well. But removing tags without visually informing the user and therefore changing the contents is in my opinion an absolute no-go. You can imagine that I had to spent half of my sunday to fix my broken mindmaps. And now I am worried how to progress and work with my mindmaps. Everytime I open them up I have to worry that my work (and this also includes the comments I made within the html sections of the richtext node) are erased.

The comments inside the html parts within the html source editor are as important data and value to me as the visual prepared html rendered results in the notes editor. Ever throught about the possibility that the comment areas are there by intend, so I can parse external sources inside the html (xml) structure of the mindmap? I know this is probably not the normal use case. But at the end it is one.

**Reply by dpolivaev** [reply]
It is not too late to consider your use case because the version removing unknown tags has not been released as stable yet. I have time for Freeplane only on my long weekends and I am going to address the problem and probably answer here.

**Comment by dpolivaev** [comment]
In preview 1.11.5_17 I reverted the removal of the unknown tags and also keep all comments. Please test this version properly because it has quite many changes underneath in simplyhtml component.

**Reply by dpolivaev** [reply]
Now removal of unknown tags happens only if a user sets a preference `HTML->Html Import->Save Compact Hypertext` and it is no longer the default behavior.

**Comment by Patronos** [comment]
> Please test this version properly because it has quite many changes underneath in simplyhtml component.

Hi! Just wanted to report back that from my first "normal" doings, all tags remain as they used to be. So far I don't see any changes compared to previous versions that I was using. If I find something, I'll report back.

And thank you very much.

:)
